cmd/zdb/zdb.c

```diff
@@ -936,23 +943,42 @@ dump_metaslab(metaslab_t *msp)
 
 	ASSERT(msp->ms_size == (1ULL << vd->vdev_ms_shift));
 	dump_spacemap(spa->spa_meta_objset, msp->ms_sm);
+
+	if (spa_feature_is_active(spa, SPA_FEATURE_LOG_SPACEMAP)) {
+		(void) printf("\tFlush data:\n\tunflushed txg=%llu\n\n",
+		    (u_longlong_t)metaslab_unflushed_txg(msp));
+	}
 }
 
 static void
 print_vdev_metaslab_header(vdev_t *vd)
 {
 	vdev_alloc_bias_t alloc_bias = vd->vdev_alloc_bias;
-	const char *bias_str;
+	const char *bias_str =
+	    (alloc_bias == VDEV_BIAS_LOG || vd->vdev_islog) ?
+	    VDEV_ALLOC_BIAS_LOG : (alloc_bias == VDEV_BIAS_SPECIAL) ?
+	    VDEV_ALLOC_BIAS_SPECIAL : (alloc_bias == VDEV_BIAS_DEDUP) ?
+	    VDEV_ALLOC_BIAS_DEDUP : vd->vdev_islog ? "log" : "";
+
```

Comment:
This is a very compact way to write it, but it's also a very complicated and hard to follow.  While I like the ternary operator a lot, I also don't like having more than two uses of it in an expression.  This would be easier as a bunch of if/else statements.  Personal opinion there.

Reply:
That block of code was already there (I just inlined it), but I definitely see what you mean. I'll rewrite it to use if/else.

Update: It's good that you brought that up because there is a dead path in there :-)
In the last (innermost) ternary operator we have `vd->vdev_islog ? "log" : "";`. There is no way for this expression to ever return `"log"` because the outermost ternary takes care of the `vd->vdev_islog` case. -> `(alloc_bias == VDEV_BIAS_LOG || vd->vdev_islog) ? VDEV_ALLOC_BIAS_LOG`. BTW the end result would be the same because `VDEV_ALLOC_BIAS_LOG` is `"log"`, but I wanted to bring this up because it proves the point of your suggestion in rewriting this.

Sneak preview:
```
const char *bias_str = "";
if (allo_bias == VDEV_BIAS_LOG || vd->vdev_islog) {
    bias_str = VDEV_ALLOC_BIAS_LOG;
} else if (alloc_bias == VDEV_BIAS_SPECIAL) {
    bias_str = VDEV_ALLOC_BIAS_SPECIAL;
} else if (alloc_bias == VDEV_BIAS_DEDUP) {
    bias_str = VDEV_ALLOC_BIAS_DEDUP;
}
```

Reply:
That block of code was already there (I just inlined it), but I definitely see what you mean. I'll rewrite it to use if/else.

Update: It's good that you brought that up because there is a dead path in there :-)
In the last (innermost) ternary operator we have `vd->vdev_islog ? "log" : "";`. There is no way for this expression to ever return `"log"` because the outermost ternary takes care of the `vd->vdev_islog` case. -> `(alloc_bias == VDEV_BIAS_LOG || vd->vdev_islog) ? VDEV_ALLOC_BIAS_LOG`. BTW the end result would be the same because `VDEV_ALLOC_BIAS_LOG` is `"log"`, but I wanted to bring this up because it proves the point of your suggestion in rewriting this.

Sneak preview:
```
const char *bias_str = "";
if (allo_bias == VDEV_BIAS_LOG || vd->vdev_islog) {
    bias_str = VDEV_ALLOC_BIAS_LOG;
} else if (alloc_bias == VDEV_BIAS_SPECIAL) {
    bias_str = VDEV_ALLOC_BIAS_SPECIAL;
} else if (alloc_bias == VDEV_BIAS_DEDUP) {
    bias_str = VDEV_ALLOC_BIAS_DEDUP;
}
```